### PR TITLE
Added ChatAction

### PR DESCRIPTION
**Pull request by grisstyl**
I added a ChatAction which supports four variables. Might need more but these are what I thought of.

### CODE REVIEW

**Comment by NathanWolf** [comment]
Cool, thanks!

A couple of requests:

1. Don't return null, I think SpellResult.PLAYER_REQUIRED is perfect for that case.
2. I'm not sure that the color translation needs to be optional? It's a nice feature, but pretty much all of the text in the result of the plugin assumes color translation with the "&" character. If you want to keep it, what I would suggest is getting rid of the "translate_alt_color_codes" flag and just having it skip translation if you blank out the "alt_color_code" parameter... but I think my inclination would be to just remove them. Up to you!

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+            p.chat(message);
+        } else {
+            if (translateAltColorCodes) {
+                for (int i = 0; i < randomizedMessages.size(); i++) {
+                    randomizedMessages.set(i, ChatColor.translateAlternateColorCodes(altColorCode, randomizedMessages.get(i)));
+                }
+            }
+
+            p.chat(randomizedMessages.get(ThreadLocalRandom.current().nextInt(randomizedMessages.size())));
```
Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: Use `CastContext#getRandom`

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    @Override
+    public SpellResult perform(CastContext context) {
+        if (!(context.getTargetEntity() instanceof Player)) return null;
+
+        Player p = ((Player) context.getTargetEntity());
+
+        if (randomizedMessages == null || randomizedMessages.isEmpty()) {
+            if (message == null) return SpellResult.FAIL;
+
+            if (translateAltColorCodes) message = ChatColor.translateAlternateColorCodes(altColorCode, message);
```
Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: You should do this translation at load time in `#prepare(CastContext,ConfigurationSection)`

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+            p.chat(message);
+        } else {
+            if (translateAltColorCodes) {
+                for (int i = 0; i < randomizedMessages.size(); i++) {
+                    randomizedMessages.set(i, ChatColor.translateAlternateColorCodes(altColorCode, randomizedMessages.get(i)));
```
Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: Also do this at load time

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    public void prepare(CastContext context, ConfigurationSection parameters) {
+        super.prepare(context, parameters);
+
+        this.message = parameters.getString("message");
+        this.randomizedMessages = parameters.getStringList("randomized_messages");
+        this.translateAltColorCodes = parameters.getBoolean("translate_alt_color_codes", true);
+
+        // Uhh, this is kind of terrible, but there doesn't appear to be a FileConfiguration#getChar or similar
+
+        this.altColorCode = parameters.getString("alt_color_code", "&").charAt(0);
```
Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: When is this useful? Can't we just let the user do a find and replace themselves if they have it in a different format?

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    @Override
+    public SpellResult perform(CastContext context) {
+        if (!(context.getTargetEntity() instanceof Player)) return null;
+
+        Player p = ((Player) context.getTargetEntity());
+
+        if (randomizedMessages == null || randomizedMessages.isEmpty()) {
```
Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: You could translate single messages into a singleton list, to avoid unnecessary code duplication here.

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+    private char altColorCode;
+
+    @Override
+    public void prepare(CastContext context, ConfigurationSection parameters) {
+        super.prepare(context, parameters);
+
+        this.message = parameters.getString("message");
+        this.randomizedMessages = parameters.getStringList("randomized_messages");
+        this.translateAltColorCodes = parameters.getBoolean("translate_alt_color_codes", true);
```
Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: Same as below.

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+    private boolean translateAltColorCodes;
+
+    private char altColorCode;
+
+    @Override
+    public void prepare(CastContext context, ConfigurationSection parameters) {
+        super.prepare(context, parameters);
+
+        this.message = parameters.getString("message");
```
Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: It would probably be useful if this can reference messages defined in the `messages.yml` file.
So messages can be translated / changed more easily.

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+import java.util.List;
+import java.util.concurrent.ThreadLocalRandom;
+
+/**
+ * Copyright Tyler Grissom 2018
+ */
+public class ChatAction extends BaseSpellAction {
+
+    private String message;
+
```
Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: No need to put a blank line between every property.
Blank lines can be used to logically group properties.

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+import com.elmakers.mine.bukkit.api.spell.SpellResult;
+import org.bukkit.ChatColor;
+import org.bukkit.configuration.ConfigurationSection;
+import org.bukkit.entity.Player;
+
+import java.util.List;
+import java.util.concurrent.ThreadLocalRandom;
+
+/**
+ * Copyright Tyler Grissom 2018
```
Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: Don't put authorship comments in.

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java (hunk 1 above)
Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: Consider extracting local variables for better readability.

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java (hunk 4 above)
Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: This will throw an exception on empty strings.
You should make a utility method and move it to ConfigurationUtils.
(If we don't remove the alt color code parameter altogether)

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java [hunk 10]
```diff
@@ -0,0 +1,73 @@
+package com.elmakers.mine.bukkit.action.builtin;
+
+import com.elmakers.mine.bukkit.action.BaseSpellAction;
```
Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: Correct me if I'm wrong, but this import order doesn't look correct.
Does this pass checkstlye? (/ A maven build)

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java (hunk 9 above)
Comment (MineStein) [reply Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: Forgot to remove it! Again, comes with my IntelliJ setup. Thank you for reminding me!

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java (hunk 1 above)
Comment (MineStein) [reply Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: I had no idea that CastContext exposed a Random variable. Will this offer any buffs over using ThreadLocalRandom?

Implemented in the next commit.

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java (hunk 2 above)
Comment (MineStein) [reply Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: Great idea, don't know why I didn't do this before. 👍

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java (hunk 4 above)
Comment (MineStein) [reply Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: I'm sure there's some edge case for it, but you would be correct in that we can just let them do it manually.

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java (hunk 5 above)
Comment (MineStein) [reply Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: I think I've devised a better way to do this, yeah.

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java (hunk 6 above)
Comment (MineStein) [reply Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: Not sure to what this is referring, but I implemented messages.yml support and those two unnecessary fields have been removed.

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java (hunk 8 above)
Comment (MineStein) [reply Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: I usually group fields together by their type and since none of these variables shared a type I spaced them out; however, in the new commit there is a single variable as opposed to many.

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java (hunk 1 above)
Comment (MineStein) [reply Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: The new code will be much more readable.

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java (hunk 4 above)
Comment (MineStein) [reply Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: Removed this parameter.

**Comment by MineStein** [comment]
I addressed everything that @killme and @NathanWolf said above in this new commit. The code is much less verbose and I removed the unnecessary variables from before.

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java (hunk 10 above)
Comment (killme) [reply Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: Make sure to try running maven from the command line and see if it succeeds.
We have configured a few maven plugins for consistent styling.

Import order goes
- java(x)
- org
- others

IIRC. You can check out other files as well.

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+    @Override
+    public void prepare(CastContext context, ConfigurationSection parameters) {
+        super.prepare(context, parameters);
+
+        this.message = context.getMessage(parameters.getString("message"), translate(parameters.getString("message")));
+
+        List<String> l = parameters.getStringList("randomized_messages");
+
+        if (!l.isEmpty()) this.message = translate(l.get(context.getRandom().nextInt(l.size())));
```
Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: It's probably more sensible to move the randomization to the perform method.

Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: You should also use `#getMessage` here

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+import com.elmakers.mine.bukkit.api.spell.SpellResult;
+import org.bukkit.ChatColor;
+import org.bukkit.configuration.ConfigurationSection;
+import org.bukkit.entity.Player;
+
+import java.util.List;
+
+public class ChatAction extends BaseSpellAction {
+
+    private String message;
```
Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: Could you annotate this with `@Nonnull` btw

Comment (MineStein) [reply Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: Which portion?

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java (hunk 11 above)
Comment (MineStein) [reply Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: For which? getMessage returns a String and wouldn't support the randomization, would it?

Comment (MineStein) [reply Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: Why?

Comment (MineStein) [reply Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: Also, I can't reply to the above two for some reason.

I've used this Action a lot from my MagicAddons project. It's useful for making Babbling Spells and the like, especially when designing Harry Potter spells.

I'm working remotely on a connection that blocks almost all ports without a maven installation so I'll do it when I get to my desktop

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java (hunk 12 above)
Comment (killme) [reply Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: `private @Nonnull String message;`

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java (hunk 11 above)
Comment (killme) [reply Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: It will do lookups in the messages file when you specify an individual message, but not when you specify a list.

Comment (killme) [reply Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: > Why?
As it is more of a "performing the action" thing rather than "loading the config".
I don't this makes a big difference rn, but we might change the semantics slightly in the future, and then it would only ever print the same message rather than randomizing.

> I've used this Action a lot from my MagicAddons project. It's useful for making Babbling Spells and the like, especially when designing Harry Potter spells.
Ah ok, seems like a reasonable use case.
(Might be a bit spammy, but I guess we can just leave it up to the makers of the spells)

> I'm working remotely on a connection that blocks almost all ports without a maven installation so I'll do it when I get to my desktop
Apart from downloading dependencies via http(s), you shouldn't need any other connections.

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java (hunk 12 above)
Comment (MineStein) [reply Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: Got it

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java (hunk 11 above)
Comment (MineStein) [reply Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: I'm saying I don't have a Maven installation and a ton of ports and sites are blocked on this network so I'll have to do that when I'm home.

**Comment by NathanWolf** [comment]
I guess you should both let me know when you think this is OK to pull :)

**Comment by grisstyl** [comment]
Pushing a commit to include the Nonnull annotation on the message field. I haven't ran the Maven build yet to make sure but I'll be home soon to do so.

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java (hunk 11 above)
Comment (grisstyl) [reply Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: I'm still a little confused as to what you're suggesting here.

Comment (NathanWolf) [reply Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ChatAction.java]: I think he means to use Messages to get messages in all cases.

I kind of agree, though I think I set a bad precedent with the Message action.

Messages does actually support randomized messages, too:

https://github.com/elBukkit/MagicPlugin/blob/master/MagicAPI/src/main/java/com/elmakers/mine/bukkit/api/magic/Messages.java#L15

That will return a random string from a list of strings defined in messages.yml

The idea with messages.yml is to keep ALL text in there, if possible, so translations are easier.

I kind of failed on that with the Message action, but that could be cleaned up.

Basically what this would mean for this action is that the parameters are not strings to send in chat, but message keys to look up via context.getMessages().get() or .getRandomized().

**Comment by grisstyl** [comment]
Unless there are any more changes @killme or @NathanWolf would like to propose, I pushed the build that passes the checkstyle.

**Comment by killme** [comment]
Will do some final clean ups and put this in master.

**Comment by killme** [comment]
Landed in 65a3a37
